File: packages/java/src/test/java/com/ailabtools/sdk/ClientTest.java
package com.ailabtools.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.ailabtools.sdk.exception.ApiException;
import com.ailabtools.sdk.generated.params.CommonQueryCreditsParams;
import com.ailabtools.sdk.generated.params.CutoutHdHumanBodyBackgroundRemovalParams;
import com.ailabtools.sdk.generated.params.CutoutUniversalBackgroundRemovalParams;
import com.ailabtools.sdk.generated.params.ImageLosslessEnlargementParams;
import com.ailabtools.sdk.generated.params.ImageRemoveObjectsAdvancedParams;
import com.ailabtools.sdk.generated.params.PortraitFaceAnalyzerAdvancedParams;
import com.ailabtools.sdk.generated.params.PortraitHairstyleEditingPremiumParams;
import com.ailabtools.sdk.generated.params.PortraitAIBreastExpansionParams;
import com.ailabtools.sdk.generated.params.PortraitTryOnClothesPremiumParams;
import com.ailabtools.sdk.http.FileInput;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

final class ClientTest {
    private HttpServer server;

    @AfterEach
    void stopServer() {
        if (server != null) server.stop(0);
    }

    @Test
    void premiumHairstyleRequiresStyleOrTemplate() {
        IllegalArgumentException error = assertThrows(
                IllegalArgumentException.class,
                () -> PortraitHairstyleEditingPremiumParams.builder()
                        .image(FileInput.fromBytes(new byte[] {1}, "photo.jpg"))
                        .build());

        assertEquals("At least one of hairStyle, imageTemplate is required", error.getMessage());
    }

    @Test
    void newRequiredParametersAreEnforced() {
        assertThrows(NullPointerException.class,
                () -> CutoutHdHumanBodyBackgroundRemovalParams.builder().build());
        assertThrows(NullPointerException.class,
                () -> PortraitAIBreastExpansionParams.builder().build());
        assertThrows(NullPointerException.class,
                () -> PortraitTryOnClothesPremiumParams.builder()
                        .taskType("async")
                        .personImage(FileInput.fromBytes(new byte[] {1}, "person.jpg"))
                        .build());
    }

    @Test
    void newNestedResponsesDeserialize() throws Exception {
        start(exchange -> {
            if (exchange.getRequestURI().getPath().contains("hd-portrait-background-removal")) {
                respond(exchange, 200,
                        "{\"error_code\":0,\"data\":{\"elements\":[{\"image_url\":\"https://example.com/cutout.png\"}]}}");
            } else {
                respond(exchange, 200,
                        "{\"error_code\":0,\"data\":{\"image\":\"https://example.com/result.png\"}}");
            }
        });
        AILabClient client = client();
        FileInput image = FileInput.fromBytes(new byte[] {1}, "image.jpg");

        var cutout = client.cutout().cutoutHdHumanBodyBackgroundRemoval(
                CutoutHdHumanBodyBackgroundRemovalParams.builder().image(image).build());
        var breast = client.portrait().portraitAIBreastExpansion(
                PortraitAIBreastExpansionParams.builder().personImage(image).build());

        assertEquals("https://example.com/cutout.png",
                cutout.getData().getElements().get(0).getImageUrl());
        assertEquals("https://example.com/result.png", breast.getData().getImage());
    }

    @Test
    void backgroundAliasUploadsFileAndMapsResponse() throws Exception {
        AtomicReference<String> body = new AtomicReference<>();
        AtomicReference<String> apiKey = new AtomicReference<>();
        start(exchange -> {
            body.set(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8));
            apiKey.set(exchange.getRequestHeaders().getFirst("ailabapi-api-key"));
            assertEquals("/api/cutout/general/universal-background-removal", exchange.getRequestURI().getPath());
            respond(exchange, 200, "{\"error_code\":0,\"request_id\":\"request-1\",\"data\":{\"image_url\":\"https://example.com/result.png\"}}");
        });
        AILabClient client = client();

        var result = client.background().remove(
                CutoutUniversalBackgroundRemovalParams.builder()
                        .image(FileInput.fromBytes("image".getBytes(StandardCharsets.UTF_8), "photo.jpg", "image/jpeg"))
                        .returnForm("whiteBK")
                        .build());

        assertEquals("request-1", result.getRequestId());
        assertEquals("https://example.com/result.png", result.getData().getImageUrl());
        assertEquals("test-key", apiKey.get());
        assertTrue(body.get().contains("name=\"return_form\""));
        assertTrue(body.get().contains("whiteBK"));
        assertTrue(body.get().contains("filename=\"photo.jpg\""));
    }

    @Test
    void upscaleMapsCamelCaseFields() throws Exception {
        AtomicReference<String> body = new AtomicReference<>();
        start(exchange -> {
            body.set(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8));
            respond(exchange, 200, "{\"error_code\":0,\"data\":{\"url\":\"https://example.com/upscaled.jpg\"}}");
        });

        var result = client().image().upscale(
                ImageLosslessEnlargementParams.builder()
                        .image(FileInput.fromBytes(new byte[] {1}, "photo.jpg"))
                        .upscaleFactor(4)
                        .outputFormat("jpg")
                        .outputQuality(80)
                        .build());

        assertEquals("https://example.com/upscaled.jpg", result.getData().getUrl());
        assertTrue(body.get().contains("name=\"upscale_factor\""));
        assertTrue(body.get().contains("name=\"output_format\""));
        assertTrue(!body.get().contains("name=\"upscaleFactor\""));
    }

    @Test
    void apiErrorsExposeRequestAndLogIds() throws Exception {
        start(exchange -> respond(exchange, 400,
                "{\"error_code\":1001,\"request_id\":\"request-2\",\"log_id\":\"log-2\","
                        + "\"error_detail\":{\"message\":\"Unsupported image format\"}}"));

        ApiException error = assertThrows(ApiException.class,
                () -> client().common().commonQueryCredits(CommonQueryCreditsParams.builder().build()));

        assertEquals(400, error.getStatusCode());
        assertEquals("request-2", error.getRequestId());
        assertEquals("log-2", error.getLogId());
        assertEquals("Unsupported image format", error.getMessage());
    }

    @Test
    void typedArrayResponsesDeserialize() throws Exception {
        start(exchange -> {
            if (exchange.getRequestURI().getPath().contains("face-analyzer")) {
                respond(exchange, 200,
                        "{\"error_code\":0,\"data\":{\"pupils\":[1,2.5],\"gender_list\":[0,1]}}");
            } else {
                respond(exchange, 200,
                        "{\"error_code\":0,\"data\":{\"binary_data_base64\":[\"first\",\"second\"]}}");
            }
        });
        AILabClient client = client();
        FileInput image = FileInput.fromBytes(new byte[] {1}, "image.jpg");

        var analysis = client.portrait().portraitFaceAnalyzerAdvanced(
                PortraitFaceAnalyzerAdvancedParams.builder().image(image).build());
        var removal = client.image().imageRemoveObjectsAdvanced(
                ImageRemoveObjectsAdvancedParams.builder().image(image).mask(image).build());

        assertEquals(Arrays.asList(1.0, 2.5), analysis.getData().getPupils());
        assertEquals(Arrays.asList(0, 1), analysis.getData().getGenderList());
        assertEquals(Arrays.asList("first", "second"), removal.getData().getBinaryDataBase64());
    }

    @Test
    void waitForTaskPollsUntilSuccess() throws Exception {
        AtomicInteger calls = new AtomicInteger();
        start(exchange -> {
            int status = calls.incrementAndGet() == 1 ? 1 : 2;
            assertEquals("task_id=task-1", exchange.getRequestURI().getQuery());
            respond(exchange, 200, "{\"error_code\":0,\"data\":{\"task_status\":" + status + "}}");
        });

        var result = client().waitForTask(
                "task-1", Duration.ZERO, Duration.ofSeconds(1), true);

        assertEquals(2, result.getData().getTaskStatus());
        assertEquals(2, calls.get());
    }

    private AILabClient client() {
        return AILabClient.builder("test-key")
                .baseUrl("http://127.0.0.1:" + server.getAddress().getPort())
                .timeout(Duration.ofSeconds(5))
                .build();
    }

    private void start(ThrowingHandler handler) throws IOException {
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", exchange -> {
            try {
                handler.handle(exchange);
            } catch (Exception error) {
                respond(exchange, 500, "{\"error_code\":1,\"error_msg\":\"test handler failed\"}");
            }
        });
        server.start();
    }

    private static void respond(HttpExchange exchange, int status, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.close();
    }

    @FunctionalInterface
    private interface ThrowingHandler {
        void handle(HttpExchange exchange) throws Exception;
    }
}

File: packages/kotlin/src/test/kotlin/com/ailabtools/sdk/ClientTest.kt
package com.ailabtools.sdk

import com.ailabtools.sdk.generated.params.CommonQueryCreditsParams
import com.ailabtools.sdk.generated.params.ImageLosslessEnlargementParams
import com.ailabtools.sdk.generated.params.PortraitFaceAnalyzerAdvancedParams
import com.ailabtools.sdk.generated.params.PortraitHairstyleEditingPremiumParams
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.runBlocking
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

internal class ClientTest {
    private var server: MockWebServer? = null

    @AfterEach
    fun stopServer() {
        server?.shutdown()
    }

    @Test
    fun `required-any-of validation rejects incomplete parameters`() {
        val error = assertThrows(IllegalArgumentException::class.java) {
            PortraitHairstyleEditingPremiumParams(
                image = FileInput.fromBytes(byteArrayOf(1), "photo.jpg"),
            )
        }

        assertEquals(
            "At least one of hairStyle, imageTemplate is required",
            error.message,
        )
    }

    @Test
    fun `upscale alias maps camel case fields to multipart wire names`() = runBlocking {
        val server = start()
        server.enqueue(success("""{"url":"https://example.com/upscaled.jpg"}"""))
        val client = client(server)

        val response = client.image.upscale(
            ImageLosslessEnlargementParams(
                image = FileInput.fromBytes(
                    "image".toByteArray(),
                    "photo.jpg",
                    "image/jpeg",
                ),
                upscaleFactor = 4,
                outputFormat = "jpg",
                outputQuality = 80,
            ),
        )
        val request = server.takeRequest(1, TimeUnit.SECONDS)!!
        val body = request.body.readUtf8()

        assertEquals("https://example.com/upscaled.jpg", response.data?.url)
        assertEquals("test-key", request.getHeader("ailabapi-api-key"))
        assertTrue(body.contains("name=\"upscale_factor\""))
        assertTrue(body.contains("name=\"output_format\""))
        assertTrue(body.contains("filename=\"photo.jpg\""))
        assertFalse(body.contains("name=\"upscaleFactor\""))
    }

    @Test
    fun `typed numeric arrays deserialize`() = runBlocking {
        val server = start()
        server.enqueue(
            success(
                """{"pupils":[1,2.5],"gender_list":[0,1],"face_count":1}""",
            ),
        )

        val response = client(server).portrait.portraitFaceAnalyzerAdvanced(
            PortraitFaceAnalyzerAdvancedParams(
                image = FileInput.fromBytes(byteArrayOf(1), "face.jpg"),
            ),
        )

        assertEquals(listOf(1.0, 2.5), response.data?.pupils)
        assertEquals(listOf(0, 1), response.data?.genderList)
        assertEquals(1, response.data?.faceCount)
    }

    @Test
    fun `API errors expose request and log identifiers`() {
        val server = start()
        server.enqueue(
            MockResponse()
                .setResponseCode(400)
                .setBody(
                    """{"error_code":1001,"request_id":"request-2","log_id":"log-2","error_detail":{"message":"Unsupported image format"}}""",
                ),
        )

        val error = assertThrows(ApiException::class.java) {
            runBlocking {
                client(server).common.commonQueryCredits(CommonQueryCreditsParams())
            }
        }

        assertEquals(400, error.statusCode)
        assertEquals(1001, error.errorCode)
        assertEquals("request-2", error.requestId)
        assertEquals("log-2", error.logId)
        assertEquals("Unsupported image format", error.message)
    }

    @Test
    fun `waitForTask polls until success`() = runBlocking {
        val server = start()
        server.enqueue(success("""{"task_status":1}"""))
        server.enqueue(success("""{"task_status":2}"""))

        val result = client(server).waitForTask(
            taskId = "task-1",
            intervalMillis = 0,
            timeoutMillis = 1_000,
        )

        assertEquals(2, result.data?.taskStatus)
        assertEquals("task_id=task-1", server.takeRequest().requestUrl?.query)
        assertEquals("task_id=task-1", server.takeRequest().requestUrl?.query)
    }

    @Test
    fun `invalid API key is rejected locally`() {
        assertThrows(IllegalArgumentException::class.java) {
            AILabClient(" ")
        }
    }

    @Test
    fun `client timeout overrides shorter injected OkHttp timeouts`() = runBlocking {
        val server = start()
        server.enqueue(
            success("[]").setBodyDelay(50, TimeUnit.MILLISECONDS),
        )
        val restrictiveClient = OkHttpClient.Builder()
            .readTimeout(1, TimeUnit.MILLISECONDS)
            .build()
        val client = AILabClient(
            apiKey = "test-key",
            baseUrl = server.url("/").toString(),
            timeoutMillis = 1_000,
            httpClient = restrictiveClient,
        )

        val response = client.common.commonQueryCredits(CommonQueryCreditsParams())

        assertEquals(0, response.errorCode)
    }

    private fun start(): MockWebServer = MockWebServer().also {
        it.start()
        server = it
    }

    private fun client(server: MockWebServer): AILabClient =
        AILabClient(
            apiKey = "test-key",
            baseUrl = server.url("/").toString(),
            timeoutMillis = 5_000,
        )

    private fun success(data: String): MockResponse =
        MockResponse().setBody("""{"error_code":0,"data":$data}""")
}
